File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
def dp():
    nk = list(map(int, input().split()))
    dp = [100001] * (nk[1] + 1)

    if nk[0] >= nk[1]:
        return nk[0] - nk[1]

    for i in range(nk[0]):
        dp[i] = nk[0] - i

    dp[nk[0]] = 0

    for i in range(nk[0] + 1, nk[1] + 1):
        if i % 2 == 0:
            dp[i] = min(dp[i // 2] + 1, dp[i - 1] + 1)
        else:
            dp[i] = min(dp[i - 1] + 1, dp[(i + 1) // 2] + 2)

    return dp[nk[1]]

print(dp())

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.js
class Queue {
  constructor() {
    this.items = {};
    this.headIndex = 0;
    this.tailIndex = 0;
  }
  enqueue(item) {
    this.items[this.tailIndex] = item;
    this.tailIndex++;
  }
  dequeue() {
    const item = this.items[this.headIndex];
    delete this.items[this.headIndex];
    this.headIndex++;
    return item;
  }
  peek() {
    return this.items[this.headIndex];
  }
  getLength() {
    return this.tailIndex - this.headIndex;
  }
}

let fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().split("\n");

const [n, k] = input[0].split(" ").map(Number);
let field = Array.from({ length: 100001 }, () => -1);

let q = new Queue();
q.enqueue(n);
field[n] = 0;

while (field[k] == -1) {
  let cur = q.dequeue();
  let nexts = [];
  nexts.push(cur - 1);
  nexts.push(cur + 1);
  nexts.push(cur * 2);

  for (next of nexts) {
    if (next < 0 || next > 100001) continue;
    if (field[next] == -1) {
      q.enqueue(next);
      field[next] = field[cur] + 1;
    }
  }
}

console.log(field[k]);
